circe-display-images.el

```diff
@@ -0,0 +1,197 @@
+;; (enable-circe-display-images)
+
+;;; Code:
+
+(require 'circe)
+(require 'url)
+
+;;;###autoload
+(defun enable-circe-display-images ()
+  "Enable the Display Images module for Circe.
+This module displays various image types when they are linked in a channel"
+  (interactive)
+  (dolist (buf (buffer-list))
+    (with-current-buffer buf
+      (when (eq major-mode 'circe-channel-mode)
+        (add-circe-display-images))))
+  (add-hook 'circe-channel-mode-hook
+            'add-circe-display-images))
+
+(defun disable-circe-display-images ()
+  "Disable the Display Images module for Circe.
+See `enable-circe-display-images'."
+  (interactive)
+  (dolist (buf (buffer-list))
+    (with-current-buffer buf
+      (when (eq major-mode 'circe-channel-mode)
+        (remove-circe-display-images))))
+  (remove-hook 'circe-channel-mode-hook
+               'add-circe-display-images))
+
+(defun add-circe-display-images ()
+  "Add `circe-display-images' to `lui-pre-output-hook'."
+  (add-hook 'lui-pre-output-hook 'circe-display-images))
+
+(defun remove-circe-display-images ()
+  "Remove `circe-display-images' from `lui-pre-output-hook'."
+  (remove-hook 'lui-pre-output-hook 'circe-display-images))
+
+(defgroup circe-display-images nil
+  "Image display properties for Circe"
+  :prefix "circe-display-images"
+  :group 'circe)
+
+(defcustom circe-display-images-image-regex
+  "\\(https?:\/\/.*?\.\\\(?:png\\|jpg\\|jpeg\\|svg\\|gif\\)\\)"
+  "Regex used to find images in channel messages. This regex needs to be
+greedy to match multiple images on the same line."
+  :group 'circe-display-images)
+
+(defcustom circe-display-images-max-height 400
+  "The image's maximum allowed height. Images will be scaled down if they
+are larger than this"
+  :group 'circe-display-images)
+
+(defcustom circe-display-images-background nil
+  "Background used for the images background, if image supports transparency.
+Defaults to the frame's background color."
+  :group 'circe-display-images)
+
+(defcustom circe-display-images-animate-gifs nil
+  "Animate any gifs that are displayed. This might slow down Emacs."
+  :group 'circe-display-images)
+
+(defvar-local circe-display-images-text-property-map (make-hash-table
+                                                      :test 'equal)
+  "A hash map used to manage display transitions.
+
+The keys are urls, and the values are a plist with an `:image-property', and a
+`:display-image-p'. `:image-property' is the display property of the image, and
+`:display-image-p' is a flag telling us whether the image is currently visible
+or not. This map serves to keep track of display transitions, and as a mapping
+between the URL and its downloaded image.
+
+Unfortunately we can't map from URL to the image position in the buffer
+because 1) the lui library can move text around when executing the
+`lui-post-output-hooks' and 2) as we toggle images, that also changes other
+images' position in the buffer.")
```

Comment:
This strikes me as odd. Wouldn't it be sufficient to store all relevant information in text properties? Transitions could then just be changing the values of relevant text properties.

Reply:
The problem I ran into is that the process of changing text properties is done with:
```lisp
(add-text-properties
           left current-pos (list from nil to current-image-property)
```
which requires the position. And since the position of an image is volatile, I had to match against the actual image property. But maybe I'm misunderstanding your suggestion?

Reply:
Hm. Maybe I am. I never used images in Emacs.